src/source/tile.js

```diff
@@ -195,18 +203,61 @@ class Tile {
     }
 
     setExpiryData(data) {
-        if (data.cacheControl) this.cacheControl = data.cacheControl;
-        if (data.expires) this.expires = data.expires;
+        const prior = this.expires;
+
+        if (data.cacheControl) {
+            const parsedCC = util.parseCacheControl(data.cacheControl);
+            if (parsedCC['max-age']) this.expires = this.timeAdded + parsedCC['max-age'] * 1000;
+        } else if (data.expires) {
+            this.expires = new Date(data.expires).getTime();
+        }
+
+        if (this.expires) {
+            const now = Date.now();
+
+            if (this.expires > now) {
+                this.expired = false;
+            } else if (!prior) {
+                this.expired = true;
+            } else if (this.expires < prior) {
+                // Expiring date is going backwards:
+                // fall back to exponential backoff
+                this.expired = true;
+
+            } else {
+                const delta = this.expires - prior;
+
+                if (!delta) {
+                    // Server is serving the same expired resource over and over: fall
+                    // back to exponential backoff.
+                    this.expired = true;
+
+                } else {
+                    // Assume that either the client or the server clock is wrong and
+                    // try to interpolate a valid expiration date (from the client POV)
+                    // observing a minimum timeout.
+                    this.expires = now + Math.max(delta, CLOCK_SKEW_RETRY_TIMEOUT);
+
+                }
+            }
+
+            if (this.expired) {
```

Comment:
Am I reading it right that  `this.expired` only used within this method?  If so, I think it'd be better as a local var.